src/tqec/sketchup/model.py

```diff
@@ -0,0 +1,511 @@
+from __future__ import annotations
+
+import typing as ty
+from dataclasses import dataclass
+import pathlib
+
+import collada
+import networkx as nx
+import numpy as np
+import numpy.typing as npt
+
+from tqec.sketchup.block import Face, FaceType, load_library_blocks, BlockType
+from tqec.exceptions import TQECException
+
+LIGHT_RGBA = (1, 1, 1, 1)
+DARK_RGBA = (0.1176470588235294, 0.1176470588235294, 0.1176470588235294, 1)
+YELLOW_RGBA = (1, 1, 0.396078431372549, 1)
+
+ASSET_AUTHOR = "TQEC Community"
+ASSET_AUTHORING_TOOL_TQEC = "TQEC Python Package"
+ASSET_AUTHORING_TOOL_SKETCHUP = "SketchUp 8.0.15158"
+ASSET_UNIT_NAME = "inch"
+ASSET_UNIT_METER = 0.02539999969303608
+
+MATERIAL_SYMBOL = "MaterialSymbol"
+
+
+@dataclass(frozen=True)
+class InstancePosition:
+    x: float
+    y: float
+    z: float
+
+    def offset_by(self, dx: float, dy: float, dz: float) -> InstancePosition:
+        return InstancePosition(self.x + dx, self.y + dy, self.z + dz)
+
+    def to_tuple(self) -> tuple[float, float, float]:
+        return self.x, self.y, self.z
+
+    def norm2(self) -> float:
+        return self.x**2 + self.y**2 + self.z**2
+
+    def __eq__(self, other: InstancePosition) -> bool:
+        return isinstance(other, InstancePosition) and all(
+            abs(i - j) < 1e-12 for i, j in zip(self.to_tuple(), other.to_tuple())
+        )
+
+
+@dataclass(frozen=True)
+class BlockInstance:
+    id: int
+    block_type: BlockType
+    position: InstancePosition
+    scale: float = 1.0
+
+
+class SketchUpModel:
+    def __init__(self) -> None:
+        self._instance_graph = nx.DiGraph()
+        # A monotonic increasing id for instances
+        self._instance_id = 0
+        self._connector_to_endpoints: dict[int, tuple[int, int]] = {}
+        self._root: int | None = None
+
+    @property
+    def num_cubes(self) -> int:
+        return self._instance_graph.number_of_nodes()
+
+    @property
+    def num_connectors(self) -> int:
+        return self._instance_graph.number_of_edges()
+
+    @property
+    def num_instances(self) -> int:
+        return self.num_cubes + self.num_connectors
```

Comment:
In practice, this `num_instances` property and the `_instance_id` one can be used interchangeably right?

Reply:
It is for now. But if we want to add functionality to remove instances, the two might not be equal since `_instance_id` is monotonically increasing by design.